### PR TITLE
feat: Add a RateCounter JavaScript Class which can be used standalone for counting and rate calculations

### PR DESCRIPTION
**Pull request by JakeChampion**
Each commit here should hopefully show the steps required for exposing a new host-based piece of functionality via a JavaScript Class.

- Added WIT (WebAssembly Interface Types) definitions for the Fastly Edge Rate Limiter functionality that is now available from the host.
- Generated the code bindings for this new functionality using `make regenerate-world`
- Also brought in the functionality via WASM_IMPORT calls (this is because we are not yet using WASM Components)
- Added implementations for the new definitions which came from `make regenerate-world`, that internally use of the ABI implementations
- Created C++ wrappers for the rate-counter functionality, to make is nicer to work with within our code-base
- Implemented a JavaScript Class named RateCounter, which has methods for all the rate-counter functionality the host provides
- Created a JavaScript module named `fastly:edge-rater-limiter` which exports the RateCounter Class, this is how applications will import the functionality
- Added documentation to the website for the RateCounter Class
- Added tests which use the JavaScript RateCounter Class to confirm that it works correctly

Example application using the new functionality:
```js
import { RateCounter } from 'fastly:edge-rate-limiter';

addEventListener("fetch", event => {
  const rc = new RateCounter('rc')

  const entry = event.client.address;

  rc.increment(entry, 1);

  event.respondWith(new Response(`rate for ${entry} over 1 second is ${rc.lookupRate(entry, 1)}
  count for ${entry} over 10 seconds is ${rc.lookupCount(entry, 10)}`
  ))
})

```

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -10,7 +10,7 @@ defaults:
   run:
     shell: bash
 env:
-  viceroy_version: 0.8.1
+  viceroy_version: 0.9.4
```
Comment (JakeChampion) [inline .github/workflows/main.yml:13]: This is needed as the new erl features are in this new version of viceroy

**Review by elliottt — APPROVED** [review]
Looks good to me! I had a few suggestions, but none of them block merging this.

File: runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+    return false;
+  }
+
+  // Convert window parameter into a number
+  double window;
+  if (!JS::ToNumber(cx, args.get(1), &window)) {
+    return false;
+  }
+
+  if (window != 1 && window != 10 && window != 60) {
```
[the remark this answers is not in the document]
Comment (JakeChampion) [reply runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp:76]: Yes, this one is validated on the compute platform (not yet validated on viceroy)

File: runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+  MOZ_ASSERT(JS::GetReservedSlot(self, Slots::Name).isString());
+  JS::RootedString name_val(cx, JS::GetReservedSlot(self, Slots::Name).toString());
+  auto name = core::encode(cx, name_val);
+  if (!name) {
+    return false;
+  }
+
+  auto rc = host_api::RateCounter(name);
+
+  auto res = rc.lookup_rate(entry, window);
```
[the remark this answers is not in the document]
Comment (JakeChampion) [reply runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp:90]:
```suggestion
  auto res = host_api::RateCounter(name).lookup_rate(entry, window);
```

File: runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+  }
+
+  MOZ_ASSERT(JS::GetReservedSlot(self, Slots::Name).isString());
+  JS::RootedString name_val(cx, JS::GetReservedSlot(self, Slots::Name).toString());
+  auto name = core::encode(cx, name_val);
+  if (!name) {
+    return false;
+  }
+
+  auto res = host_api::RateCounter(name).rc.increment(entry, delta);
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp:49]:
```suggestion
  auto res = host_api::RateCounter(name).increment(entry, delta);
```

File: runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+  MOZ_ASSERT(JS::GetReservedSlot(self, Slots::Name).isString());
+  JS::RootedString name_val(cx, JS::GetReservedSlot(self, Slots::Name).toString());
+  auto name = core::encode(cx, name_val);
+  if (!name) {
+    return false;
+  }
+
+  auto rc = host_api::RateCounter(name);
+
+  auto res = rc.lookup_count(entry, duration);
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp:134]:
```suggestion
  auto rc = host_api::RateCounter(name).lookup_count(entry, duration);
```

File: types/edge-rate-limiter.d.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    /**
+     * Lookup the current rate for entry for a given window
+     * @param entry The entry to lookup
+     * @param window The window to lookup alongside the entry, has to be either 1 second, 10 seconds, or 60 seconds
+     */
+    lookupRate(entry: string, window: 1 | 10 | 60): number;
+    /**
+     * Lookup the current count for entry for a given duration
+     * @param entry The entry to lookup
+     * @param duration The duration to lookup alongside the entry, has to be either, 10, 20, 30, 40, 50, or 60 seconds.
```
[the remark this answers is not in the document]
Comment (JakeChampion) [reply types/edge-rate-limiter.d.ts:27]: We've gone with `duration` as that is what the Rust and Go SDKs are using:
https://docs.rs/fastly/latest/fastly/erl/struct.RateCounter.html#method.lookup_count
https://pkg.go.dev/github.com/fastly/compute-sdk-go@v1.3.0/erl#RateCounter.LookupCount

File: runtime/js-compute-runtime/host_interface/host_api.h [hunk 7]
```diff
@@ -780,6 +780,18 @@ class Backend final {
...
+
+class RateCounter final {
+public:
+  std::string_view name;
+  RateCounter() = default;
+  explicit RateCounter(std::string_view name) : name{name} {}
+
+  Result<Void> increment(std::string_view entry, uint32_t delta);
+  Result<uint32_t> lookup_rate(std::string_view entry, uint32_t window);
+  Result<uint32_t> lookup_count(std::string_view entry, uint32_t duration);
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/host_interface/host_api.h:792]:
```suggestion
  static Result<Void> increment(std::string_view name, std::string_view entry, uint32_t delta);
  static Result<uint32_t> lookup_rate(std::string_view name, std::string_view entry, uint32_t window);
  static Result<uint32_t> lookup_count(std::string_view name, std::string_view entry, uint32_t duration);
```

File: runtime/js-compute-runtime/host_interface/host_api.cpp [hunk 8]
```diff
@@ -1791,4 +1791,50 @@ Result<BackendHealth> Backend::health(std::string_view name) {
   return res;
 }
 
+Result<Void> RateCounter::increment(std::string_view entry, uint32_t delta) {
+  auto name_str = string_view_to_world_string(this->name);
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/host_interface/host_api.cpp:1795]:
```suggestion
Result<Void> RateCounter::increment(std::string_view name, std::string_view entry, uint32_t delta) {
  auto name_str = string_view_to_world_string(name);
```

File: runtime/js-compute-runtime/host_interface/host_api.cpp [hunk 9]
```diff
@@ -1791,4 +1791,50 @@ Result<BackendHealth> Backend::health(std::string_view name) {
...
+    return Result<Void>::err(err);
+  }
+
+  return Result<Void>::ok();
+}
+
+Result<uint32_t> RateCounter::lookup_rate(std::string_view entry, uint32_t window) {
+  Result<uint32_t> res;
+
+  auto name_str = string_view_to_world_string(this->name);
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/host_interface/host_api.cpp:1809]:
```suggestion
Result<uint32_t> RateCounter::lookup_rate(std::string_view name, std::string_view entry, uint32_t window) {
  Result<uint32_t> res;

  auto name_str = string_view_to_world_string(name);
```

File: runtime/js-compute-runtime/host_interface/host_api.cpp [hunk 10]
```diff
@@ -1791,4 +1791,50 @@ Result<BackendHealth> Backend::health(std::string_view name) {
...
+    res.emplace(ret);
+  }
+
+  return res;
+}
+
+Result<uint32_t> RateCounter::lookup_count(std::string_view entry, uint32_t duration) {
+  Result<uint32_t> res;
+
+  auto name_str = string_view_to_world_string(this->name);
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/host_interface/host_api.cpp:1826]:
```suggestion
Result<uint32_t> RateCounter::lookup_count(std::string_view name, std::string_view entry, uint32_t duration) {
  Result<uint32_t> res;

  auto name_str = string_view_to_world_string(name);
```

File: runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  MOZ_ASSERT(JS::GetReservedSlot(self, Slots::Name).isString());
+  JS::RootedString name_val(cx, JS::GetReservedSlot(self, Slots::Name).toString());
+  auto name = core::encode(cx, name_val);
+  if (!name) {
+    return false;
+  }
+
+  auto res = host_api::RateCounter(name).lookup_count(entry, duration);
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp:130]:
```suggestion
  auto res = host_api::RateCounter::lookup_count(entry, duration);
```

File: runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  MOZ_ASSERT(JS::GetReservedSlot(self, Slots::Name).isString());
+  JS::RootedString name_val(cx, JS::GetReservedSlot(self, Slots::Name).toString());
+  auto name = core::encode(cx, name_val);
+  if (!name) {
+    return false;
+  }
+
+  auto res = host_api::RateCounter(name).lookup_rate(entry, window);
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp:88]:
```suggestion
  auto res = host_api::RateCounter::lookup_rate(name, entry, window);
```

File: runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+  }
+
+  MOZ_ASSERT(JS::GetReservedSlot(self, Slots::Name).isString());
+  JS::RootedString name_val(cx, JS::GetReservedSlot(self, Slots::Name).toString());
+  auto name = core::encode(cx, name_val);
+  if (!name) {
+    return false;
+  }
+
+  auto res = host_api::RateCounter(name).increment(entry, delta);
```
Comment (JakeChampion) [inline runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp:49]:
```suggestion
  auto res = host_api::RateCounter::increment(name, entry, delta);
```

File: runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+  if (!JS::ToNumber(cx, args.get(1), &duration)) {
+    return false;
+  }
+
+  if (duration != 10 && duration != 20 && duration != 30 && duration != 40 && duration != 50 &&
+      duration != 60) {
+    JS_ReportErrorASCII(
+        cx, "lookupCount: duration parameter must be either: 10, 20, 30, 40, 50, or 60");
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (JakeChampion) [reply runtime/js-compute-runtime/builtins/edge-rate-limiter.cpp:121]: Same for this one, it is validated on Compute and not yet in viceroy

**Pull request merged**